### PR TITLE
[TEP-0135] coschedule isolate pipelinerun

### PR DESCRIPTION
**Pull request by QuanZhang-William**
# Changes
Part of [#6740]. [TEP-0135][tep-0135] introduces a feature that allows a cluster operator to ensure that all of a PipelineRun's pods are scheduled to the same node.

This commit implements `coschedule: isolate-pipelinerun` coschedule mode by modifying [PodAntiAffinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity) terms in the `Affinity Assistant StatefulSets`, which enforces only 1 pipelinerun is running in a node at the same time.

/kind feature

[#6740]: #6740
[tep-0135]: https://github.com/tektoncd/community/blob/main/teps/0135-coscheduling-pipelinerun-pods.md

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant.go:317]: What's the reason for this change?

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant.go:361]: why is this no longer using the label selector? Is the reason because the label is not accurate for coschedule = pipelineruns?

**Review by lbernick — COMMENTED** [review]
Thanks Quan! Can you include a release note for this feature and update the documentation to explain that it is implemented?

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:317]: The function now needs `aaBehavior` to determine the anti-pod affinity terms

**Comment by QuanZhang-William** [comment]
> Thanks Quan! Can you include a release note for this feature and update the documentation to explain that it is implemented?

Hi Lee! This is actually not the end to end support for `coschedule-isolate-pipelinerun` mode, so I don't this is the right time to update docs and release note.

Once we merged this PR, https://github.com/tektoncd/pipeline/pull/6927 will have e2e support for all 3 modes, and we can update doc and release note in that one instead.

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:361]: oh sorry-- I just didn't read this carefully enough. It might be worth adding a quick comment about why it's required for isolate pipelineruns but optional for the other methods.

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/affinity_assistant.go:346]:
```suggestion
	repelOtherAffinityAssistantsPodAffinityTerm := corev1.PodAffinityTerm{
```
(nit)

**Comment by QuanZhang-William** [comment]
> > > Thanks Quan! Can you include a release note for this feature and update the documentation to explain that it is implemented?
> > 
> > 
> > Hi Lee! This is actually not the end to end support for `coschedule-isolate-pipelinerun` mode, so I don't this is the right time to update docs and release note.
> > Once we merged this PR, #6927 will have e2e support for all 3 modes, and we can update doc and release note in that one instead.
> 
> Oh I see, thanks. Can you update the description of #6927 (or add a comment) to explain that it depends on this PR? It's probably worth putting a hold as well.

Comments added in #6927 (and put it on hold)

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:317]: I prefer the previous behavior. The new behavior adds more logic to tests, making them more brittle and makes `createOrUpdateAffinityAssistant` and `affinityAssistantStatefulSet` less readable by adding a new argument. (If you do decide to keep the new logic, please add the new argument to the function docstring.)

**Review by lbernick — APPROVED** [review]

File: pkg/reconciler/pipelinerun/affinity_assistant.go
Comment (QuanZhang-William) [reply pkg/reconciler/pipelinerun/affinity_assistant.go:317]: SGTM, I have reverted to the previous behavior!

And also a somewhat related fix: https://github.com/tektoncd/pipeline/pull/6930 🙏

**Review by Yongxuanzhang — COMMENTED** [review]
/lgtm

**Pull request merged**
